### PR TITLE
Changes to support velero 1.10

### LINKED ISSUE #485: Update doc and code regarding restic
**Describe the bug**
In the upcoming Velero v1.10 along with the introduction to support kopia, some of the CLI parameters, resources are renamed:
Design doc: https://github.com/vmware-tanzu/velero/blob/main/design/unified-repo-and-kopia-integration/unified-repo-and-kopia-integration.md

More details in the code change:
https://github.com/vmware-tanzu/velero/pull/5370
https://github.com/vmware-tanzu/velero/pull/5390
https://github.com/vmware-tanzu/velero/pull/5432

We need to update the following doc section and test it:
https://github.com/vmware-tanzu/velero-plugin-for-vsphere/blob/main/docs/vanilla.md#backup-vsphere-cns-file-volumes

Command to install Velero with restic is changed:
https://github.com/vmware-tanzu/velero/blob/main/design/unified-repo-and-kopia-integration/unified-repo-and-kopia-integration.md#installation

**To Reproduce**

[Steps to reproduce the behavior]

**Expected behavior**

[A clear and concise description of what you expected to happen]

**Troubleshooting Information**

[Please refer to the [Troubleshooting](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/blob/main/docs/troubleshooting.md) page and collect the required information]

**Screenshots**

[If applicable, add screenshots to help explain your problem.]

**Anything else you would like to add:**

[Miscellaneous information that will assist in solving the issue]

### PR DESCRIPTION
**Pull request by deepakkinni**
**What this PR does / why we need it**:
Changes to support velero 1.10

Updated to go 1.18

**Testing**
Build/Unit
```
make push:

harbor-repo.vmware.com/velero/data-manager-for-plugin:vel_1_10_v1-c8d90ee-08.Dec.2022.11.49.13
harbor-repo.vmware.com/velero/velero-plugin-for-vsphere:vel_1_10_v1-c8d90ee-08.Dec.2022.11.49.13
harbor-repo.vmware.com/velero/backup-driver:vel_1_10_v1-c8d90ee-08.Dec.2022.11.49.13

make test:

Running tests: ./pkg/...
go test ./pkg/... -timeout=300s
Success!
```

Backup:

Without specifying default-volumes-to-fs-backup with annotation on the Pod
```
dkinni@dkinniCMD6R snapshot-vanilla-testing % velero backup create dec21bk4 --include-namespaces=foo --snapshot-volumes   
Backup request "dec21bk4" submitted successfully.
Run `velero backup describe dec21bk4` or `velero backup logs dec21bk4` for more details.

dkinni@dkinniCMD6R snapshot-vanilla-testing % velero backup describe dec21bk4                                          
Name:         dec21bk4
Namespace:    velero
Labels:       velero.io/storage-location=default
Annotations:  velero.io/source-cluster-k8s-gitversion=v1.25.2
              velero.io/source-cluster-k8s-major-version=1
              velero.io/source-cluster-k8s-minor-version=25

Phase:  Completed


Logs:
time="2022-12-22T04:41:37Z" level=info msg="Executing custom action" backup=velero/dec21bk4 logSource="pkg/backup/item_backupper.go:321" name=example-vanilla-block-pvc namespace=foo resource=persistentvolumeclaims
time="2022-12-22T04:41:37Z" level=info msg="Backing up resource persistentvolumeclaims: blocked = false" backup=velero/dec21bk4 cmd=/plugins/velero-plugin-for-vsphere logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/plugin/backup_pvc_action_plugin.go:54" pluginName=velero-plugin-for-vsphere
time="2022-12-22T04:41:37Z" level=info msg="VSphere PVCBackupItemAction for PVC foo/example-vanilla-block-pvc started. Storage Class Name: example-vanilla-block-sc" backup=velero/dec21bk4 cmd=/plugins/velero-plugin-for-vsphere logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/plugin/backup_pvc_action_plugin.go:70" pluginName=velero-plugin-for-vsphere
time="2022-12-22T04:41:37Z" level=info msg="Fetching underlying PV for PVC foo/example-vanilla-block-pvc" backup=velero/dec21bk4 cmd=/plugins/velero-plugin-for-vsphere logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/plugin/backup_pvc_action_plugin.go:96" pluginName=velero-plugin-for-vsphere
time="2022-12-22T04:41:37Z" level=info msg="Skipping PVC foo/example-vanilla-block-pvc, PV pvc-01490ccc-792a-4332-88b6-3f984581d609 will be backed up using fs backup" backup=velero/dec21bk4 cmd=/plugins/velero-plugin-for-vsphere logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/plugin/backup_pvc_action_plugin.go:109" pluginName=velero-plugin-for-vsphere
time="2022-12-22T04:41:37Z" level=info msg="VSphere PVCBackupItemAction for PVC foo/example-vanilla-block-pvc completed with err: <nil>" backup=velero/dec21bk4 cmd=/plugins/velero-plugin-for-vsphere logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/plugin/backup_pvc_action_plugin.go:73" pluginName=velero-plugin-for-vsphere
```

With specifying default-volumes-to-fs-backup and no annotation on the Pod
```
dkinni@dkinniCMD6R snapshot-vanilla-testing % velero backup create dec21bk5 --include-namespaces=foo --snapshot-volumes --default-volumes-to-fs-backup
Backup request "dec21bk5" submitted successfully.
Run `velero backup describe dec21bk5` or `velero backup logs dec21bk5` for more details.
dkinni@dkinniCMD6R snapshot-vanilla-testing % velero backup describe dec21bk5
Name:         dec21bk5
Namespace:    velero
Labels:       velero.io/storage-location=default
Annotations:  velero.io/source-cluster-k8s-gitversion=v1.25.2
              velero.io/source-cluster-k8s-major-version=1
              velero.io/source-cluster-k8s-minor-version=25

Phase:  Completed

Logs:
time="2022-12-22T04:46:13Z" level=info msg="Executing custom action" backup=velero/dec21bk5 logSource="pkg/backup/item_backupper.go:321" name=example-vanilla-block-pvc namespace=foo resource=persistentvolumeclaims
time="2022-12-22T04:46:13Z" level=info msg="Backing up resource persistentvolumeclaims: blocked = false" backup=velero/dec21bk5 cmd=/plugins/velero-plugin-for-vsphere logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/plugin/backup_pvc_action_plugin.go:54" pluginName=velero-plugin-for-vsphere
time="2022-12-22T04:46:13Z" level=info msg="VSphere PVCBackupItemAction for PVC foo/example-vanilla-block-pvc started. Storage Class Name: example-vanilla-block-sc" backup=velero/dec21bk5 cmd=/plugins/velero-plugin-for-vsphere logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/plugin/backup_pvc_action_plugin.go:70" pluginName=velero-plugin-for-vsphere
time="2022-12-22T04:46:13Z" level=info msg="Fetching underlying PV for PVC foo/example-vanilla-block-pvc" backup=velero/dec21bk5 cmd=/plugins/velero-plugin-for-vsphere logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/plugin/backup_pvc_action_plugin.go:96" pluginName=velero-plugin-for-vsphere
time="2022-12-22T04:46:13Z" level=info msg="Skipping PVC foo/example-vanilla-block-pvc, PV pvc-01490ccc-792a-4332-88b6-3f984581d609 will be backed up using fs backup" backup=velero/dec21bk5 cmd=/plugins/velero-plugin-for-vsphere logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/plugin/backup_pvc_action_plugin.go:109" pluginName=velero-plugin-for-vsphere
time="2022-12-22T04:46:13Z" level=info msg="VSphere PVCBackupItemAction for PVC foo/example-vanilla-block-pvc completed with err: <nil>" backup=velero/dec21bk5 cmd=/plugins/velero-plugin-for-vsphere logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/plugin/backup_pvc_action_plugin.go:73" pluginName=velero-plugin-for-vsphere
```

No annotation on the pod and without specifying default-volumes-to-fs-backup
```
This is regular backup, and works as expected.
```

**Which issue(s) this PR fixes**:
Fixes #485

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Changes to support velero v1.10
```
Signed-off-by: Deepak Kinni <dkinni@vmware.com>

### CODE REVIEW

File: pkg/builder/object_meta.go [hunk 1]
```diff
@@ -91,14 +91,6 @@ func setMapEntries(m map[string]string, vals ...string) map[string]string {
...
 }
 
-// WithClusterName is a functional option that applies the specified
-// cluster name to an object.
-func WithClusterName(val string) func(obj metav1.Object) {
-	return func(obj metav1.Object) {
-		obj.SetClusterName(val)
-	}
-}
-
```
Comment (xing-yang) [inline pkg/builder/object_meta.go:101]: Is this function not used before?

File: go.mod [hunk 2]
```diff
@@ -1,148 +1,173 @@
...
-	sigs.k8s.io/controller-runtime v0.10.2 // indirect
-	sigs.k8s.io/structured-merge-diff/v4 v4.1.2 // indirect
+	k8s.io/kube-openapi v0.0.0-20220614142933-1062c7ade5f8 // indirect
+	sigs.k8s.io/controller-runtime v0.12.2 // indirect
+	sigs.k8s.io/json v0.0.0-20220525155127-227cbc7cc124 // indirect
+	sigs.k8s.io/structured-merge-diff/v4 v4.2.1 // indirect
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
 
 replace github.com/gogo/protobuf => github.com/gogo/protobuf v1.3.2
```
Comment (xing-yang) [inline go.mod:167]: Can you get rid of the "replace" if possible?  I missed it last time.

Comment (deepakkinni) [reply go.mod]: done

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/772/

**Review by xinyanw409 — COMMENTED** [review]
Do we need to push vendor/ folder to the repo?

**Comment by deepakkinni** [comment]
> Do we need to push vendor/ folder to the repo?

It was not a part of .gitignore in the past, so I assume it's expected.

**Comment by xing-yang** [comment]
>> Do we need to push vendor/ folder to the repo?

> It was not a part of .gitignore in the past, so I assume it's expected.

We don't have a vendor folder.  I don't think we should start having this in the github.

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/773/

**Comment by xing-yang** [comment]
Please also update the document:
https://github.com/vmware-tanzu/velero-plugin-for-vsphere/blob/main/docs/vanilla.md#backup-vsphere-cns-file-volumes

File: pkg/plugin/backup_pvc_action_plugin.go [hunk 3]
```diff
@@ -100,13 +100,13 @@ func (p *NewPVCBackupItemAction) Execute(item runtime.Unstructured, backup *vele
...
-	isResticUsed, err := pluginUtil.IsPVCBackedUpByRestic(pvc.Namespace, pvc.Name, kubeClient.CoreV1(), boolptr.IsSetToTrue(backup.Spec.DefaultVolumesToRestic))
+	// Do nothing if fs backup is used to backup this PV
+	isFsBackupUsed, err := pluginUtil.IsPVCBackedUpByFsBackup(pvc.Namespace, pvc.Name, kubeClient.CoreV1(), boolptr.IsSetToTrue(backup.Spec.DefaultVolumesToFsBackup))
 	if err != nil {
 		return nil, nil, errors.WithStack(err)
 	}
-	if isResticUsed {
-		p.Log.Infof("Skipping PVC %s/%s, PV %s will be backed up using restic", pvc.Namespace, pvc.Name, pv.Name)
+	if isFsBackupUsed {
+		p.Log.Infof("Skipping PVC %s/%s, PV %s will be backed up using fs backup", pvc.Namespace, pvc.Name, pv.Name)
```
[the remark this answers is not in the document]
Comment (deepakkinni) [reply pkg/plugin/backup_pvc_action_plugin.go:109]: Update the testing. It works as expected.

**Comment by deepakkinni** [comment]
> Please also update the document: https://github.com/vmware-tanzu/velero-plugin-for-vsphere/blob/main/docs/vanilla.md#backup-vsphere-cns-file-volumes

Done.

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/775/

**Comment by xing-yang** [comment]
@deepakkinni With or without specifying default-volumes-to-fs-backup, the logs look identical.  So what is the difference there?  What does `default-volumes-to-fs-backup` do?

**Comment by deepakkinni** [comment]
> @deepakkinni With or without specifying default-volumes-to-fs-backup, the logs look identical. So what is the difference there? What does `default-volumes-to-fs-backup` do?

From the plugin's perspective, the primary change is `DefaultVolumesToRestic` param is deprecated, and instead, users are supposed to specify `DefaultVolumesToFsBackup`/`default-volumes-to-fs-backup`.

The primary goal of the restic support is to ensure that specific volumes are skipped from backup. This is done by annotating the pod (https://velero.io/docs/v1.10/file-system-backup/#to-back-up).

When `--default-volumes-to-fs-backup` is `False` the plugin looks for `backup.velero.io/backup-volumes` annotation and skips them.

When `--default-volumes-to-fs-backup` is `True`, the plugin automatically assumes that all volumes attached to the pod are to be backed up by fs backup and hence are skipped.

**Comment by xing-yang** [comment]
Can you add one test:
- Remove annotations + without default flag
- Remove annotations + add the default flag

**Comment by deepakkinni** [comment]
> Can you add one test:
> 
>     * Remove annotations + without default flag
> 
>     * Remove annotations + add the default flag

Clarified this in testing.

**Review by xing-yang — APPROVED** [review]
LGTM

**Pull request merged**